### PR TITLE
fix if-then-else, anyOf, dependencies

### PR DESCRIPTION
**Pull request by karenetheridge**
### Summary
fixes bugs: #190, #196, #192 where various keywords were not fully supported

### Motivation
see their respective issues for background and references

I am not quite sure what error messages are best: "Missing property" was already being used by the 'required' keyword, and it's not great to have two different errors sharing the same message, so I made the 'required' error "Missing required property", so 'dependencies' can have "Missing dependent property". However, changing an existing error message might break unit tests in downstream code.

### CODE REVIEW

File: lib/JSON/Validator/Error.pm [hunk 1]
```diff
@@ -34,7 +34,8 @@ our $MESSAGES = {
     additionalProperties => 'Properties not allowed: %3.',
     maxProperties        => 'Too many properties: %3/%4.',
     minProperties        => 'Not enough properties: %3/%4.',
-    required             => 'Missing property.',
+    required             => 'Missing required property.',
+    dependencies         => 'Missing dependent property.',
```
Comment (jhthorsen) [inline lib/JSON/Validator/Error.pm]: I think this is incomplete: It should mention the path/name of the dependent property. How about something like "%3 missing dependent property." or something? If you can make it shorter or rephrase it so "%3" comes at the end, then that would be super.

File: lib/JSON/Validator.pm [hunk 2]
```diff
@@ -551,15 +549,25 @@ sub _validate {
...
   }
-  elsif ($rules = $schema->{anyOf}) {
+
+  if (my $rules = $schema->{anyOf}) {
     push @errors,
       $self->_validate_any_of($to_json ? $$to_json : $_[1], $path, $rules);
   }
-  elsif ($rules = $schema->{oneOf}) {
+
+  if (my $rules = $schema->{oneOf}) {
```
[the remark this answers is not in the document]
Comment (karenetheridge) [reply lib/JSON/Validator.pm]: a combination of oneOf, anyOf and allOf all at the same time? hey, you never know :D

File: lib/JSON/Validator.pm [hunk 3]
```diff
@@ -900,6 +902,22 @@ sub _validate_type_object {
...
+        push @errors, E json_pointer($path, $dep_key),
+          [object => dependencies => $data_key];
+      }
+    }
+    elsif (ref $schema->{dependencies}{$data_key} eq 'HASH') {
+      push @errors,
+        $self->_validate_type_object($data, $path,
+        $schema->{dependencies}{$data_key});
+    }
+  }
```
Comment (jhthorsen) [inline lib/JSON/Validator.pm]: Sorry for going back/forward on this code, but I think this is easier to read:

```perl
  my $dependencies = $schema->{dependencies} || {};
  for my $k (keys %$dependencies) {
    next if not exists $data->{$k};

    if (ref $dependencies->{$k} eq 'ARRAY') {
      push @errors,
        map { E json_pointer($path, $_), [object => dependencies => $k]; }
        grep { !exists $data->{$_} } @{$dependencies->{$k}};
    }
    elsif (ref $dependencies->{$k} eq 'HASH') {
      push @errors,
        $self->_validate_type_object($data, $path, $dependencies->{$k});
    }
  }
```

File: lib/JSON/Validator/Error.pm [hunk 4]
```diff
@@ -35,6 +35,7 @@ our $MESSAGES = {
     maxProperties        => 'Too many properties: %3/%4.',
     minProperties        => 'Not enough properties: %3/%4.',
     required             => 'Missing property.',
+    dependencies         => 'Missing property: Dependee: %3.',
```
Comment (jhthorsen) [inline lib/JSON/Validator/Error.pm]: Too many colons. How about "Missing property. Dependee: %3."

File: lib/JSON/Validator/Error.pm (hunk 1 above)
Comment (jhthorsen) [reply lib/JSON/Validator/Error.pm]: See https://github.com/mojolicious/json-validator/pull/197#discussion_r379681124

File: lib/JSON/Validator/Error.pm (hunk 4 above)
Comment (karenetheridge) [reply lib/JSON/Validator/Error.pm]: fixed.

File: lib/JSON/Validator.pm (hunk 3 above)
Comment (karenetheridge) [reply lib/JSON/Validator.pm]: the map and grep is not easier to read.

Comment (karenetheridge) [reply lib/JSON/Validator.pm]: ..but ok.

**Comment by jhthorsen** [comment]
Awesome 👍

**Pull request merged**